src/gar.lua

```diff
@@ -2191,4 +2191,46 @@ function gar.nextRedelegationsPruneTimestamp()
 	return NextRedelegationsPruneTimestamp
 end
 
+--- @class DelegatesFromAllGateways
+--- @field cursorKey string -- delegateAddress_gatewayAddress
+--- @field address WalletAddress
+--- @field gatewayAddress WalletAddress
+--- @field startTimestamp Timestamp
+--- @field delegatedStake mARIO
+--- @field vaultedStake mARIO
+
+--- @param cursor string|nil -- cursorKey of the last item in the previous page
+--- @param limit number
+--- @param sortBy string|nil
+--- @param sortOrder string|nil
+--- @return PaginatedTable<DelegatesFromAllGateways>
+function gar.getPaginatedDelegatesFromAllGateways(cursor, limit, sortBy, sortOrder)
+	--- @type DelegatesFromAllGateways[]
+	local allDelegations = {}
+
+	for gatewayAddress, gateway in pairs(gar.getGatewaysUnsafe()) do
+		for delegateAddress, delegate in pairs(gateway.delegates) do
+			table.insert(allDelegations, {
+				cursorKey = delegateAddress .. "_" .. gatewayAddress,
+				address = delegateAddress,
+				gatewayAddress = gatewayAddress,
+				startTimestamp = delegate.startTimestamp,
+				delegatedStake = delegate.delegatedStake,
+				vaultedStake = utils.reduce(delegate.vaults, function(acc, _, vault)
+					return acc + vault.balance
+				end, 0),
+			})
+		end
+	end
```

Comment:
don't think there's any way to avoid the O(n^2) complexity here without bloating global state in other ways. Ideally we'd look to memorize or use more efficient data structure for lookups (hash tables, maps, etc.). It may be useful to add `totalVaultedStake` attribute to gateways to avoid the additional reduction, but that would take some effort. Callers will just have to face the potential they will hit memory/CPU limts on CUs that enforce lower limits.

Reply:
blue sky memory thinking over the weekend -- ideally here we'd have memoized ordered array of gateway keys and the keys would be smaller than the current addresses. so then we could paginate on gatewayKey_delegateAddress, remove sorting, and never pull down the whole gateways table only the ones we need. it is all a big lift though